diff -Nur nginx-upstream-fair-a18b4099fbd458111983200e098b6f0c8efed4bc.orig/config nginx-upstream-fair-a18b4099fbd458111983200e098b6f0c8efed4bc/config
--- nginx-upstream-fair-a18b4099fbd458111983200e098b6f0c8efed4bc.orig/config	2012-04-08 22:03:04.000000000 +0800
+++ nginx-upstream-fair-a18b4099fbd458111983200e098b6f0c8efed4bc/config	2016-11-21 16:14:49.000000000 +0800
@@ -1,3 +1,10 @@
 ngx_addon_name=ngx_http_upstream_fair_module
-HTTP_MODULES="$HTTP_MODULES ngx_http_upstream_fair_module"
-NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_http_upstream_fair_module.c"
+if test -n "$ngx_module_link"; then
+    ngx_module_type=HTTP
+    ngx_module_name=$ngx_addon_name
+    ngx_module_srcs="$ngx_addon_dir/ngx_http_upstream_fair_module.c"
+    . auto/module
+else
+    HTTP_MODULES="$HTTP_MODULES $ngx_addon_name"
+    NGX_ADDON_SRCS="$NGX_ADDON_SRCS $ngx_addon_dir/ngx_http_upstream_fair_module.c"
+fi
