diff -Nur passenger-release-5.1.12-old/build/packaging.rb passenger-release-5.1.12/build/packaging.rb
--- passenger-release-5.1.12-old/build/packaging.rb	2017-11-23 08:15:47.000000000 +0800
+++ passenger-release-5.1.12/build/packaging.rb	2017-12-21 18:10:16.000000000 +0800
@@ -115,7 +115,7 @@
 end
 
 desc "Create a fakeroot, useful for building native packages"
-task :fakeroot => [:apache2, :nginx, 'nginx:as_dynamic_module', :doc] do
+task :fakeroot => [:nginx, 'nginx:as_dynamic_module'] do
   require 'rbconfig'
   include RbConfig
 
@@ -234,8 +234,8 @@
   sh "cp src/nginx_module/* #{fake_nginx_module_source_dir}/"
 
   # Documentation
-  sh "mkdir -p #{fake_docdir}"
-  sh "cp -R doc/images #{fake_docdir}/"
+  #sh "mkdir -p #{fake_docdir}"
+  #sh "cp -R doc/images #{fake_docdir}/"
 
   # User binaries
   sh "mkdir -p #{fake_bindir}"
@@ -256,14 +256,14 @@
     if PhusionPassenger::Packaging::EXECUTABLES_WITH_FREE_RUBY.include?(exe)
       shebang = psg_free_ruby
     else
-      shebang = psg_ruby
+      shebang = psg_free_ruby
     end
     change_shebang("#{fake_sbindir}/#{exe}", shebang)
   end
 
   # Apache 2 module
-  sh "mkdir -p #{File.dirname(fake_apache2_module_path)}"
-  sh "cp #{APACHE2_TARGET} #{fake_apache2_module_path}"
+  #sh "mkdir -p #{File.dirname(fake_apache2_module_path)}"
+  #sh "cp #{APACHE2_TARGET} #{fake_apache2_module_path}"
 
   # Ruby extension sources
   sh "mkdir -p #{fake_ruby_extension_source_dir}"
diff -Nur passenger-release-5.1.12-old/gitsubmodules.sh passenger-release-5.1.12/gitsubmodules.sh
--- passenger-release-5.1.12-old/gitsubmodules.sh	1970-01-01 08:00:00.000000000 +0800
+++ passenger-release-5.1.12/gitsubmodules.sh	2017-12-21 18:05:45.000000000 +0800
@@ -0,0 +1,13 @@
+#!/bin/sh
+set -e
+sed -i -e '/packaging\/rpm/d;/_rpm_/d' \
+       -e '/packaging\/debian/d;/_apt_/d' .gitmodules || exit 0
+git init
+git config -f .gitmodules --get-regexp '^submodule\..*\.path$' |
+    while read path_key path_str
+    do
+        url_key=$(echo $path_key | sed 's/\.path/.url/')
+        url=$(git config -f .gitmodules --get "$url_key")
+        [ -d $path_str ] && rmdir $path_str || exit 0
+        git submodule add $url $path_str
+    done
